open_lm/evaluate.py

```diff
@@ -0,0 +1,191 @@
+import logging
+import math
+import time
+import copy
+
+import torch
+import torch.distributed as dist
+
+try:
+    import wandb
+except ImportError:
+    wandb = None
+
+from open_lm.data import sample_chunk
+from open_lm.distributed import is_master
+from open_lm.precision import get_autocast
+from open_lm.meters import (
+    AverageMeter,
+    ConfidenceIntervalMeter,
+    gather_meters,
+)
+
+
+@torch.inference_mode()
+def evaluate(model, data, start_epoch, args, writer):
+    """
+    evaluates perplexity on validation data
+    """
+    if is_master(args):
+        print("=> begin evaluation")
+    device = torch.device(args.device)
+    autocast = get_autocast(args.precision)
+
+    model.eval()
+
+    data["val"].set_epoch(start_epoch)  # set epoch in process safe manner via sampler or shared_epoch
+    dataloader = data["val"].dataloader
+
+    # NOTE: max_num_batches = 0 corresponds to exhausting iterator
+    max_num_batches = dataloader.num_batches
+
+    losses_m = AverageMeter()
+    batch_time_m = AverageMeter()
+    data_time_m = AverageMeter()
+    sps_m = AverageMeter()
+    spspg_m = AverageMeter()
+    losses_seq_ci_m = ConfidenceIntervalMeter()
+    losses_tok_ci_m = ConfidenceIntervalMeter()
+
+    end = time.time()
+    loss = torch.nn.CrossEntropyLoss(reduction="none")
+
+    for i, batch in enumerate(dataloader):
+        if i == max_num_batches and max_num_batches != 0:
```

Comment:
Presumably the dataloader should end exactly when this is true right? If so, I think we don't need this check. If not, we might be missing some data from eval?

Reply:
so the dataloader always tries to exhaust the data. we have a special flag `--val-num-samples` for early stoping of eval. the current convention is if max_num_batches is 0 then we exhaust the loader, but `None` makes better sense here so updating to that

Reply:
will add a comment to this effect

Reply:
ahh okay, makes sense, fine as is then